pandas/_libs/lib.pyx

```diff
@@ -2359,3 +2356,106 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
             output[i] = default
 
     return maybe_convert_objects(output)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void convert_and_set_item(object item, Py_ssize_t index,
+                                      object[:] result,
+                                      bint keep_trivial_numbers):
+    cdef:
+        bint do_convert = 1
+        double double_item
+
+    if keep_trivial_numbers:
+        if isinstance(item, int) and Py_SIZE(item) < 2:
+            if <int>item == 0:
+                do_convert = 0
+        elif isinstance(item, float):
+            double_item = item
+            if double_item == 0.0 or double_item != double_item:
+                do_convert = 0
+
+    if do_convert and not isinstance(item, (str, bytes)):
+        item = PyObject_Str(item)
+
+    result[index] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void put_object_as_unicode(list lst, Py_ssize_t idx,
+                                       object item):
+    if not isinstance(item, str):
+        item = PyObject_Str(item)
+    lst[idx] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef object _concat_date_cols(tuple date_cols,
```

Comment:
again I would encourage you to profile rather than simply re-writing everything. You have added a ton of code here, which increases the complexity for future maitainers. If you can narrow this patch down to something reasonable could consider.

Reply:
We **did** profile, we didn't optimize blindly :)
Is winning a dozen of seconds on more realistic benchmarks (like reading a 10M csv file from Mortgage dataset) not worth adding extra ~150 lines of Cython?..

Reply:
> Is winning a dozen of seconds on more realistic benchmarks (like reading a 10M csv file from Mortgage dataset) not worth adding extra ~150 lines of Cython?..

you have a ton of extra code here that I really doubt actually affects the perf. you are adding a ton of complexity.

Reply:
I'm again looking at a simple csv-reading benchmark we have shown above (compared to micro-benchmarks in `asv`).
I've added one combined datecolumn parsing to it and dropped most code that is not needed to time the running, so the code is like this:

<details>
  <summary>Reading birthdays, two date columns (one concatenated)</summary>

```python
#!/usr/bin/env python
import sys
import time
import pandas as pd

if len(sys.argv) > 1:
    dataset_file = sys.argv[1]
else:
    sys.exit("please specify dataset")

time_call = time.time()
df = pd.read_csv(dataset_file, sep=',', parse_dates=['date', ['year', 'month', 'day']])
time_call = time.time() - time_call

print("read_csv for %s took %f seconds" % (dataset_file, time_call))
```
</details>

Running it on 10 million lines of [aforementioned Birthdays.csv](https://vincentarelbundock.github.io/Rdatasets/csv/mosaicData/Birthdays.csv) on my laptop with `Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz` yields the following numbers:

branch | wallclock time
-|-
master | 53.9591 sec
patched | 41.3366 sec

This shows 12.6 seconds saved, or about 23% faster parsing.